File: opcional/aeropuerto.java
import java.util.concurrent.Semaphore;

public class Avion extends Thread {

    private int id;

    private static final int K_HANGARES = 10;
    private static final int AVIONES = 30;
    private static Semaphore hangar;
    private static Semaphore pista;

    public Avion(int id) {
        this.id = id;
    }

    public void aterrizar() {
        hangar.aquire();
        pista.aquire();

        Thread.sleep(1000);
        System.out.println("El avion " + this.id + " aterriza");

        pista.release();
    }

    public void cargar() {
        Thread.sleep(500);
        System.out.println("El avion " + this.id + " carga");
    }

    public void descargar() {
        Thread.sleep(500);
        System.out.println("El avion " + this.id + " descarga");

        pista.aquire();
        hangar.release();
    }
    public void despegar() {
        Thread.sleep(1000);
        System.out.println("El avion " + this.id + " despega");

        pista.release();
    }


    public void run() {
        aterrizar();
        cargar();
        descargar();
        despegar();
    }

    public static void main(String args[]) {
        this.hangar = new Semaphore(K_HANGARES);
        this.pista = new Semaphore(1);
        ArrayList<Thread> threads = new ArrayList<Thread>();

        for (int i = 0; i < aviones; i++) {
            Thread thr = new Avion(i);
            thr.start();
            threads.add(thr);
        }

        try{
            for (int i = 0; i < aviones; i++) {
                Thread thr = threads.get(i);
                thr.join();
            }
        }catch(InterruptedException e) {
            return;
        }
    }
}

File: opcional/aeropuerto.py
import random
import threading
import time

AVIONES = 25
K_HANGARES = 10

class Avion(threading.Thread):
    hangar = threading.Semaphore(K_HANGARES) #Semaforo que controla los hangares disponibles (inicializado a K_HANGARES)
    pista = threading.Lock()                 #Mutex que controla el acceso a la pista

    def __init__(self, id):
        super(Avion, self).__init__()
        self.id = id

    def aterrizar(self):
        Avion.hangar.acquire();     #reservamos el hangar antes de intentar reservar la pista
        print("El avion {} reserva hangar".format(self.id))
        with Avion.pista:           #con el hangar reservado, aterrizamos una vez la pista este disponible
            time.sleep(0.25);
            print("El avion {} aterriza".format(self.id))

    def cargar(self):
        time.sleep(0.1);
        print("El avion {} carga".format(self.id))

    def descargar(self):
        time.sleep(0.1);
        print("El avion {} descarga".format(self.id))
        Avion.pista.acquire()
        Avion.hangar.release()
        print("El avion {} deja el hangar".format(self.id))

    def despegar(self):
        time.sleep(0.25);
        print("El avion {} despega".format(self.id))
        Avion.pista.release()

    def run(self):
        self.aterrizar()
        self.cargar()
        self.descargar()
        self.despegar()

def main():

    aviones = []

    for i in range(AVIONES):
        aviones.append(Avion(i))

    for a in aviones:
        time.sleep(random.uniform(0.1,0.2)); #esperamos un tiempo aleatorio para la llegada de los aviones
        a.start()

    for a in aviones:
        a.join()


if __name__ == "__main__":
    main()
